pkg/controller/corecc/controller.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// logic for reading ConfigConnector (CC) and ConfigConnectorContext (CCC) CRs.
+package corecc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/k8s"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// CoreReconciler is a controller that ONLY observers the ConfigConnector (CC) or ConfigConnectorContext (CCC) CRs.
+// Only observing changes allows us to use the same reconciler structure for both and will ensure we don't step
+// onto the operator's controllers for CC & CCC.
+type CoreReconciler struct {
+	mgr                     manager.Manager
+	lc                      *LiveConfig
+	gvk                     schema.GroupVersionKind
+}
+
+var _ reconcile.Reconciler = &CoreReconciler{}
+
+func (r *CoreReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
```

Comment:
I don't think you need to do this; you should just call this at the top of each reconcile operation (probably in a helper function!)

```
var cc kcc.ConfigConnector
if err := r.client.Get(ctx, types.NamespacedName{Name: foo}, &cc); err != nil {
return fmt.Errorf("getting configconnector: %w", err)
}
if cc.Spec.Mode == "Namespaced" { // We probably have this anyway, but this is where it comes from
var ccc kcc.ConfigConnectorContext
if err := r.client.Get(ctx, types.NamespacedName{Name: foo, Namespace: req.GetNamespace()}, &ccc); err != nil {
return fmt.Errorf("getting configconnectorcontext: %w", err)
}
}
```

In a normal controller-runtime operator, that will be cached.  We turned off caching in KCC, so you'll want to turn it back on for these kinds, it should be in the manager.Options.

Reply:
thanks for the guidance! opened https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1300

the ideas in these PR are self contained enough that it doesn't make sense to me to continue on it so will close it in favor of the one above